qiskit_nature/circuit/library/initial_states/_helper.py

```diff
@@ -0,0 +1,197 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Private helper functions for initial states."""
+
+from typing import Iterable, List, Sequence, Tuple, Union
+
+import numpy as np
+
+from qiskit import QuantumRegister
+from qiskit.circuit import Gate, Qubit
+from qiskit.circuit.library import RZGate, XGate, XXPlusYYGate
+
+
+def _prepare_slater_determinant_jordan_wigner(  # pylint: disable=invalid-name
+    register: QuantumRegister, transformation_matrix: np.ndarray
+) -> Iterable[Tuple[Gate, Tuple[Qubit, ...]]]:
+    """Prepare a Slater determinant under the Jordan-Wigner Transform."""
+    m, n = transformation_matrix.shape
+
+    # set the first n_particles qubits to 1
+    for i in range(m):
+        yield XGate(), (register[i],)
+
+    # if all orbitals are filled, no further operations are needed
+    if m == n:
+        return
+
+    current_matrix = transformation_matrix
+
+    # zero out top right corner by rotating rows; this is a no-op
+    for j in reversed(range(n - m + 1, n)):
+        # Zero out entries in column j
+        for i in range(m - n + j):
+            # Zero out entry in row i if needed
+            if not np.isclose(current_matrix[i, j], 0.0):
+                givens_matrix = _givens_matrix(current_matrix[i, j], current_matrix[i + 1, j])
+                current_matrix = _apply_matrix_to_slices(current_matrix, givens_matrix, [i, i + 1])
+
+    # decompose matrix into Givens rotations
+    decomposition: List[Tuple[Gate, Tuple[Qubit, ...]]] = []
+    for i in range(m):
+        # zero out the columns in row i
+        for j in range(n - m + i, i, -1):
+            if not np.isclose(current_matrix[i, j], 0.0):
+                # compute Givens rotation
+                givens_matrix = _givens_matrix(current_matrix[i, j], current_matrix[i, j - 1])
+                theta = np.arcsin(np.real(givens_matrix[1, 0]))
+                phi = -np.angle(givens_matrix[1, 1])
+                # add operations
+                decomposition.append((RZGate(phi), (register[j - 1],)))
+                decomposition.append(
+                    (XXPlusYYGate(2 * theta, -np.pi / 2), (register[j], register[j - 1]))
+                )
+                # update matrix
+                current_matrix = _apply_matrix_to_slices(
+                    current_matrix, givens_matrix, [(Ellipsis, j), (Ellipsis, j - 1)]
+                )
+
+    yield from reversed(decomposition)
+
+
+def _prepare_fermionic_gaussian_state_jordan_wigner(  # pylint: disable=invalid-name
```

Comment:
There is quite some code overlap with the previous function. I suggest you refactor these slightly as to avoid this duplication.

Reply:
Could you be specific? Which lines are duplicated? How should they be refactored?

Reply:
`_prepare_slater_determinant_jordan_wigner` and `_prepare_fermionic_gaussian_state_jordan_wigner` share a lot of common code. This should be improved

Reply:
Sorry, I was fooled a bit by the very similar looking structure of the code but you're right, extracting methods would overcomplicate the logic flow to a point where it would be harder to follow than this.